File: practice-projects/springboot4ddd/src/main/java/com/github/microwind/springboot4ddd/infrastructure/repository/user/UserRepositoryImpl.java
package com.github.microwind.springboot4ddd.infrastructure.repository.user;

import com.github.microwind.springboot4ddd.domain.model.user.User;
import com.github.microwind.springboot4ddd.domain.page.PageRequest;
import com.github.microwind.springboot4ddd.domain.page.PageResult;
import com.github.microwind.springboot4ddd.domain.page.SortOrder;
import com.github.microwind.springboot4ddd.domain.repository.user.UserRepository;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.jdbc.core.simple.JdbcClient;
import org.springframework.stereotype.Repository;

import java.sql.Timestamp;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

/**
 * 用户仓储实现 - MySQL 数据源
 *
 * <p>通过 {@link UserDO} + {@link UserConverter} 显式分离持久化与领域模型，
 * 与 {@code OrderJdbcRepositoryImpl} 的处理方式对称。
 *
 * @author jarry
 * @since 1.0.0
 */
@Slf4j
@Repository
public class UserRepositoryImpl implements UserRepository {

    private final JdbcClient jdbcClient;

    public UserRepositoryImpl(@Qualifier("userJdbcClient") JdbcClient jdbcClient) {
        this.jdbcClient = jdbcClient;
    }

    private static UserDO mapRowToUserDO(java.sql.ResultSet rs) throws java.sql.SQLException {
        return UserDO.builder()
                .id(rs.getLong("id"))
                .name(rs.getString("name"))
                .email(rs.getString("email"))
                .phone(rs.getString("phone"))
                .wechat(getStringOrNull(rs, "wechat"))
                .address(getStringOrNull(rs, "address"))
                .createdTime(rs.getTimestamp("created_time").toLocalDateTime())
                .updatedTime(rs.getTimestamp("updated_time").toLocalDateTime())
                .build();
    }

    /** 列不存在或为 null 时返回 null，避免 SQLException: Column 'xxx' not found */
    private static String getStringOrNull(java.sql.ResultSet rs, String column) throws java.sql.SQLException {
        try {
            rs.findColumn(column);
        } catch (java.sql.SQLException notFound) {
            return null;
        }
        return rs.getString(column);
    }

    @Override
    public User save(User user) {
        UserDO userDO = UserConverter.toDO(user);
        String sql = "INSERT INTO users (name, email, phone, address, created_time, updated_time) " +
                "VALUES (?, ?, ?, ?, ?, ?)";

        LocalDateTime now = LocalDateTime.now();
        LocalDateTime createdTime = userDO.getCreatedTime() != null ? userDO.getCreatedTime() : now;
        LocalDateTime updatedTime = userDO.getUpdatedTime() != null ? userDO.getUpdatedTime() : now;

        Long generatedId = jdbcClient.sql(sql)
                .params(userDO.getName(), userDO.getEmail(), userDO.getPhone(), 
                        userDO.getAddress(), Timestamp.valueOf(createdTime), Timestamp.valueOf(updatedTime))
                .query(Long.class)
                .single();

        user.markPersisted(generatedId);
        log.info("User saved with id: {}", user.getId());
        return user;
    }

    @Override
    public Optional<User> findById(Long id) {
        String sql = "SELECT * FROM users WHERE id = ?";
        return jdbcClient.sql(sql)
                .param(id)
                .query((rs, rowNum) -> UserConverter.toModel(UserRepositoryImpl.mapRowToUserDO(rs)))
                .optional();
    }

    @Override
    public Optional<User> findByName(String name) {
        String sql = "SELECT * FROM users WHERE name = ?";
        return jdbcClient.sql(sql)
                .param(name)
                .query((rs, rowNum) -> UserConverter.toModel(UserRepositoryImpl.mapRowToUserDO(rs)))
                .optional();
    }

    @Override
    public Optional<User> findByEmail(String email) {
        String sql = "SELECT * FROM users WHERE email = ?";
        return jdbcClient.sql(sql)
                .param(email)
                .query((rs, rowNum) -> UserConverter.toModel(UserRepositoryImpl.mapRowToUserDO(rs)))
                .optional();
    }

    @Override
    public List<User> findAll() {
        String sql = "SELECT * FROM users ORDER BY created_time DESC";
        return jdbcClient.sql(sql)
                .query((rs, rowNum) -> UserConverter.toModel(UserRepositoryImpl.mapRowToUserDO(rs)))
                .list();
    }

    @Override
    public PageResult<User> findAll(PageRequest pageRequest) {
        String countSql = "SELECT COUNT(*) FROM users";
        Integer total = jdbcClient.sql(countSql)
                .query(Integer.class)
                .single();
        long totalElements = total != null ? total : 0;

        String orderBy = buildOrderByClause(pageRequest.getSorts());
        String sql = "SELECT * FROM users" + orderBy + " LIMIT ? OFFSET ?";

        List<User> records = jdbcClient.sql(sql)
                .params(pageRequest.getPageSize(), pageRequest.getOffset())
                .query((rs, rowNum) -> UserConverter.toModel(UserRepositoryImpl.mapRowToUserDO(rs)))
                .list();

        return new PageResult<>(
                records,
                totalElements,
                pageRequest.getPageNumber(),
                pageRequest.getPageSize()
        );
    }

    /**
     * 根据领域 SortOrder 列表构建 ORDER BY 子句；为空则按 created_time 倒序。
     */
    private String buildOrderByClause(List<SortOrder> sorts) {
        if (sorts == null || sorts.isEmpty()) {
            return " ORDER BY created_time DESC";
        }

        StringBuilder sb = new StringBuilder(" ORDER BY ");
        for (int i = 0; i < sorts.size(); i++) {
            if (i > 0) {
                sb.append(", ");
            }
            SortOrder order = sorts.get(i);
            sb.append(order.getProperty()).append(' ').append(order.getDirection().name());
        }
        return sb.toString();
    }

    @Override
    public User update(User user) {
        UserDO userDO = UserConverter.toDO(user);
        String sql = "UPDATE users SET name = ?, email = ?, phone = ?, address = ?, updated_time = ? " +
                "WHERE id = ?";

        int updated = jdbcClient.sql(sql)
                .params(userDO.getName(), userDO.getEmail(), userDO.getPhone(), 
                        userDO.getAddress(), Timestamp.valueOf(LocalDateTime.now()), userDO.getId())
                .update();

        if (updated == 0) {
            throw new IllegalArgumentException("User not found with id: " + userDO.getId());
        }

        log.info("User updated with id: {}", user.getId());
        return user;
    }

    @Override
    public void deleteById(Long id) {
        String sql = "DELETE FROM users WHERE id = ?";
        int deleted = jdbcClient.sql(sql)
                .param(id)
                .update();
        if (deleted == 0) {
            throw new IllegalArgumentException("User not found with id: " + id);
        }
        log.info("User deleted with id: {}", id);
    }

    @Override
    public boolean existsByName(String name) {
        String sql = "SELECT COUNT(*) FROM users WHERE name = ?";
        Integer count = jdbcClient.sql(sql)
                .param(name)
                .query(Integer.class)
                .single();
        return count != null && count > 0;
    }

    @Override
    public boolean existsByEmail(String email) {
        String sql = "SELECT COUNT(*) FROM users WHERE email = ?";
        Integer count = jdbcClient.sql(sql)
                .param(email)
                .query(Integer.class)
                .single();
        return count != null && count > 0;
    }
}

File: practice-projects/gin-ddd/internal/infrastructure/persistence/user/user_repository_impl.go
// Package user 用户仓储实现。
//
// 仓储层只读写 UserDO,通过 UserConverter 与领域聚合根互转。
package user

import (
	"context"
	"database/sql"
	"errors"

	userModel "gin-ddd/internal/domain/model/user"
	"gin-ddd/pkg/utils"

	"github.com/jmoiron/sqlx"
)

// UserRepositoryImpl 用户仓储实现(基于 MySQL)。
type UserRepositoryImpl struct {
	db *sqlx.DB
}

// NewUserRepository 构造仓储实例。
func NewUserRepository(db *sqlx.DB) *UserRepositoryImpl {
	return &UserRepositoryImpl{db: db}
}

// Create 持久化新用户,回填主键后调用 User.MarkPersisted 记录创建事件。
func (r *UserRepositoryImpl) Create(ctx context.Context, u *userModel.User) error {
	do := toDO(u)
	query := `
		INSERT INTO users (name, email, phone, address, created_time, updated_time)
		VALUES (?, ?, ?, ?, ?, ?)
	`
	utils.GetLogger().Debug("执行INSERT用户SQL: name=%s, email=%s", do.Name, do.Email)
	result, err := r.db.ExecContext(ctx, query,
		do.Name, do.Email, do.Phone, do.Address, do.CreatedTime, do.UpdatedTime)
	if err != nil {
		utils.GetLogger().Error("插入用户失败: %v, name=%s, email=%s", err, do.Name, do.Email)
		return err
	}
	id, err := result.LastInsertId()
	if err != nil {
		utils.GetLogger().Error("获取插入ID失败: %v", err)
		return err
	}
	if err := u.MarkPersisted(id); err != nil {
		utils.GetLogger().Error("用户初始化标记失败: %v, id=%d", err, id)
		return err
	}
	utils.GetLogger().Debug("用户插入成功: id=%d, name=%s", id, do.Name)
	return nil
}

// Update 更新用户。
func (r *UserRepositoryImpl) Update(ctx context.Context, u *userModel.User) error {
	do := toDO(u)
	query := `
		UPDATE users
		SET name = ?, email = ?, phone = ?, address = ?, updated_time = ?
		WHERE id = ?
	`
	utils.GetLogger().Debug("执行UPDATE用户SQL: id=%d, name=%s, email=%s", do.ID, do.Name, do.Email)
	if _, err := r.db.ExecContext(ctx, query,
		do.Name, do.Email, do.Phone, do.Address, do.UpdatedTime, do.ID); err != nil {
		utils.GetLogger().Error("更新用户失败: %v, id=%d", err, do.ID)
		return err
	}
	utils.GetLogger().Debug("用户更新成功: id=%d", do.ID)
	return nil
}

// Delete 删除用户。
func (r *UserRepositoryImpl) Delete(ctx context.Context, id int64) error {
	query := `DELETE FROM users WHERE id = ?`
	utils.GetLogger().Debug("执行DELETE用户SQL: id=%d", id)
	result, err := r.db.ExecContext(ctx, query, id)
	if err != nil {
		utils.GetLogger().Error("删除用户失败: %v, id=%d", err, id)
		return err
	}
	rowsAffected, _ := result.RowsAffected()
	if rowsAffected == 0 {
		utils.GetLogger().Warn("删除用户: 用户不存在, id=%d", id)
	} else {
		utils.GetLogger().Debug("用户删除成功: id=%d, rowsAffected=%d", id, rowsAffected)
	}
	return nil
}

// FindByID 根据 ID 查询用户。
func (r *UserRepositoryImpl) FindByID(ctx context.Context, id int64) (*userModel.User, error) {
	const query = `
		SELECT id, name, email, phone, address, created_time, updated_time
		FROM users WHERE id = ?
	`
	var do UserDO
	err := r.db.GetContext(ctx, &do, query, id)
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return nil, nil
		}
		utils.GetLogger().Error("查询用户失败: %v, id=%d", err, id)
		return nil, err
	}
	return toModel(&do), nil
}

// FindByName 根据用户名查询用户。
func (r *UserRepositoryImpl) FindByName(ctx context.Context, name string) (*userModel.User, error) {
	const query = `
		SELECT id, name, email, phone, address, created_time, updated_time
		FROM users WHERE name = ?
	`
	var do UserDO
	err := r.db.GetContext(ctx, &do, query, name)
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return nil, nil
		}
		utils.GetLogger().Error("查询用户失败: %v, name=%s", err, name)
		return nil, err
	}
	return toModel(&do), nil
}

// FindByEmail 根据邮箱查询用户。
func (r *UserRepositoryImpl) FindByEmail(ctx context.Context, email string) (*userModel.User, error) {
	const query = `
		SELECT id, name, email, phone, address, created_time, updated_time
		FROM users WHERE email = ?
	`
	var do UserDO
	err := r.db.GetContext(ctx, &do, query, email)
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return nil, nil
		}
		utils.GetLogger().Error("查询用户失败: %v, email=%s", err, email)
		return nil, err
	}
	return toModel(&do), nil
}

// FindAll 查询所有用户。
func (r *UserRepositoryImpl) FindAll(ctx context.Context) ([]*userModel.User, error) {
	const query = `
		SELECT id, name, email, phone, address, created_time, updated_time
		FROM users ORDER BY created_time DESC
	`
	var dos []*UserDO
	if err := r.db.SelectContext(ctx, &dos, query); err != nil {
		utils.GetLogger().Error("查询所有用户失败: %v", err)
		return nil, err
	}
	return toModels(dos), nil
}
